pages/blog/the-reference-rabbit-hole.md

```diff
@@ -0,0 +1,275 @@
+
+Well, there is one thing we need to remember about JSON Schema. It is primary build for validation rules and how a validator can take input data and determine whether that input is valid against the Schema. 
+
+This means, that if you have a JSON Schema using ref `$ref` such as:
+```json
+{ "$ref": "./test.json", "minLength": 7, "maxLength": 12}
+```
+and `test.json` is defined as:
+
+```json
+{"minLength": 5, "format": "email"}
+```
+
+JSON Schema draft 2019-09, assumes that the references are resolved similar to:
+```json
+{"$ref": {"minLength": 5, "format": "email"}, "minLength": 7, "maxLength": 12}
+```
+
+This is because in validation, you want to validate that the input data is valid against the referenced schema and should [**not** be seen as a kind of merging behavior](https://github.com/APIDevTools/json-schema-ref-parser/issues/145):
+```json
+{"format": "email", "minLength": 7, "maxLength": 12}
+```
+
+This behavior is different from what is assumed when using AsyncAPI, as the last option, is more aligned with expected behavior.
+
+Furthermore, now, each schema can define it's own `$schema` that they follow, instead of ONLY being available at the root... 
+
+This leaves the question, how can we make sure that we stay consistent and don't introduce more confusion into the AsyncAPI specification? This difference is what triggered the last issue in the [spec 649](https://github.com/asyncapi/spec/issues/649).
+
+## Hard to find tooling
+This leaves us with one huge deficit, that there are so many different behaviors for references that tooling mix and matches between the specifications and what they solve.
+
+One of the most used tooling for dereferencing stuff in JS, and the one we are using is from [APIDevTools called json-schema-ref-parser](https://github.com/APIDevTools/json-schema-ref-parser). We actually use this tool to ensure **ANY** encounters of `$ref` are dereferenced, so the tool has direct access to the schema, without it having to look elsewhere for it. 
+
+However, the tool starting out being build **ONLY** for dereferencing `$ref` based on the [JSON Reference specification and the JSON Pointer specification](https://github.com/APIDevTools/json-schema-ref-parser/issues/22#issuecomment-231783185).  At least it was, now it's not easy to figure it out what it is for, as [it allows extra properties](https://github.com/APIDevTools/json-schema-ref-parser/issues/232) but [$id is not taken into account](https://github.com/APIDevTools/json-schema-ref-parser/issues/136).
+
+This leaves us in a big of a struggle, as [there are not many alternatives](https://json-schema.org/implementations.html#general-processing), for JS [@hyperjump/json-schema-core](https://github.com/jdesrosiers/json-schema-core) looks promising, but there are no tooling that our [Go parser](https://github.com/asyncapi/parser-go) can use.
+
+And with no official or community tooling, we are left with having to develop it our selves to adopt the spec...
```

Comment:
Would you like to mention https://github.com/json-schema-org/community/discussions/113 ?

Reply:
Yea sure, let's do that 👍